src/c/device-can.c

```diff
@@ -230,47 +227,64 @@ static bool can_put_handler(void *impl, const devsdk_device_t *device,
 
 	can_driver *driver = (can_driver *)impl;
 	struct can_frame frame;
+	int iret = -1;
+	bool successful_put_request = false;
+	uint8_t i = 0, k = 0; 
 
-	bool successful_get_request = false;
-	uint8_t i = 0; 
+	end_dev_params *end_dev_params_ptr = (end_dev_params *)device->address;
 
 	pthread_mutex_lock(&driver->mutex);
 
-	// Set the iterator to the first array
-	iot_data_array_iter_t iter;
-	iot_data_array_iter (values[0], &iter);
-
-	iot_log_debug(driver->lc, "CAN: PUT on device \n");
-	iot_log_debug(driver->lc, "CAN: nvalues = %d\n", nvalues);
+	if(!end_dev_params_ptr->can_IsOpened)
+	{
+		iret = OpenCan(impl, device);
+		if(iret != 0)
+			return successful_put_request;
 
-	// Move the iterator and fill the msg ID filter
-	iot_data_array_iter_next (&iter);
-	frame.can_id = *((uint32_t *) iot_data_array_iter_value (&iter));
-	iot_log_debug(driver->lc, "can_id: %d\n", frame.can_id);
+		end_dev_params_ptr->can_IsOpened = true;
+	} else
+	{
+		iot_log_debug(driver->lc, "CAN already opened..\n");
+	}
 
-	// Move the iterator and fill the data length code
-	iot_data_array_iter_next (&iter);
-	frame.can_dlc = *((uint32_t *) iot_data_array_iter_value (&iter));
-	iot_log_debug(driver->lc, "can_dlc: %d\n", frame.can_dlc);
-	for (i = 0; i<frame.can_dlc; i++)
+	for (k = 0; k < nvalues; k++)
 	{
-		// Move the iterator and fill all the can data in the frame
+		// Set the iterator to the first array
+		iot_data_array_iter_t iter;
+		iot_data_array_iter (values[k], &iter);
+
+		iot_log_debug(driver->lc, "CAN: PUT on device \n");
+		iot_log_debug(driver->lc, "CAN: nvalues = %d\n", nvalues);
+
+		// Move the iterator and fill the msg ID filter
 		iot_data_array_iter_next (&iter);
-		frame.data[i] = *((uint32_t *) iot_data_array_iter_value (&iter));
-		iot_log_debug(driver->lc, "CAN: data[%d] = %d\n", i, frame.data[i]);
-	}
+		frame.can_id = *((uint32_t *) iot_data_array_iter_value (&iter));
+		iot_log_debug(driver->lc, "can_id: %d\n", frame.can_id);
 
-        // Write the can frame to the socket
-	if (write(sock_fd, &frame, sizeof(struct can_frame)) != sizeof(struct can_frame)) {
-		perror("Write");
-		return 1;
+		// Move the iterator and fill the data length code
+		iot_data_array_iter_next (&iter);
+		frame.can_dlc = *((uint32_t *) iot_data_array_iter_value (&iter));
+		iot_log_debug(driver->lc, "can_dlc: %d\n", frame.can_dlc);
+		for (i = 0; i<frame.can_dlc; i++)
+		{
+			// Move the iterator and fill all the can data in the frame
+			iot_data_array_iter_next (&iter);
+			frame.data[i] = *((uint32_t *) iot_data_array_iter_value (&iter));
+			iot_log_debug(driver->lc, "CAN: data[%d] = %d\n", i, frame.data[i]);
+		}
+
+		// Write the can frame to the socket
+		if (write(end_dev_params_ptr->sock_fd, &frame, sizeof(struct can_frame)) != sizeof(struct can_frame)) {
+			perror("Write");
+			return 1;
```

Comment:
the mutex should be unlocked before the `return` statement

Reply:
Modified the code